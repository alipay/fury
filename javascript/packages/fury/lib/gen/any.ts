/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { TypeDescription } from "../description";
import { CodecBuilder } from "./builder";
import { BaseSerializerGenerator } from "./serializer";
import { CodegenRegistry } from "./router";
import { InternalSerializerType } from "../type";
import { Scope } from "./scope";

class AnySerializerGenerator extends BaseSerializerGenerator {
  description: TypeDescription;

  constructor(description: TypeDescription, builder: CodecBuilder, scope: Scope) {
    super(description, builder, scope);
    this.description = description;
  }

  private addDep() {
    return this.scope.declare(
      "any_ser",
      this.builder.classResolver.getSerializerById(InternalSerializerType.ANY)
    );
  }

  writeStmt(accessor: string): string {
    const name = this.addDep();
    return `${name}.writeInner(${accessor})`;
  }

  readStmt(accessor: (expr: string) => string): string {
    const name = this.addDep();
    return `${name}.readInner(${accessor})`;
  }

  toReadEmbed(accessor: (expr: string) => string): string {
    const name = this.addDep();
    return accessor(`${name}.read()`);
  }

  toWriteEmbed(accessor: string): string {
    const name = this.addDep();
    return `${name}.write(${accessor})`;
  }
}

CodegenRegistry.register(InternalSerializerType.ANY, AnySerializerGenerator);
