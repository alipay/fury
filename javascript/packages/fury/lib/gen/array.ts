/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { ArrayTypeDescription, TypeDescription } from "../description";
import { CodecBuilder } from "./builder";
import { BaseSerializerGenerator } from "./serializer";
import { CodegenRegistry } from "./router";
import { InternalSerializerType } from "../type";
import { Scope } from "./scope";

class ArraySerializerGenerator extends BaseSerializerGenerator {
  description: ArrayTypeDescription;

  constructor(description: TypeDescription, builder: CodecBuilder, scope: Scope) {
    super(description, builder, scope);
    this.description = <ArrayTypeDescription>description;
  }

  private innerMeta() {
    const inner = this.description.options.inner;
    return this.builder.meta(inner);
  }

  private innerGenerator() {
    const inner = this.description.options.inner;
    const InnerGeneratorClass = CodegenRegistry.get(inner.type);
    if (!InnerGeneratorClass) {
      throw new Error(`${inner.type} generator not exists`);
    }
    return new InnerGeneratorClass(inner, this.builder, this.scope);
  }

  writeStmt(accessor: string): string {
    const innerMeta = this.innerMeta();
    const innerGenerator = this.innerGenerator();
    const item = this.scope.uniqueName("item");
    return `
            ${this.builder.writer.varUInt32(`${accessor}.length`)}
            ${this.builder.writer.reserve(`${innerMeta.fixedSize} * ${accessor}.length`)};
            for (const ${item} of ${accessor}) {
                ${innerGenerator.toWriteEmbed(item)}
            }
        `;
  }

  readStmt(accessor: (expr: string) => string): string {
    const innerGenerator = this.innerGenerator();
    const result = this.scope.uniqueName("result");
    const len = this.scope.uniqueName("len");
    const idx = this.scope.uniqueName("idx");

    return `
            const ${len} = ${this.builder.reader.varUInt32()};
            const ${result} = new Array(${len});
            ${this.pushReadRefStmt(result)}
            for (let ${idx} = 0; ${idx} < ${len}; ${idx}++) {
                ${innerGenerator.toReadEmbed(x => `${result}[${idx}] = ${x};`)}
            }
            ${accessor(result)}
         `;
  }
}

CodegenRegistry.register(InternalSerializerType.ARRAY, ArraySerializerGenerator);
