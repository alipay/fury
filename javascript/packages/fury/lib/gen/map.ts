/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { MapTypeDescription, TypeDescription } from "../description";
import { CodecBuilder } from "./builder";
import { BaseSerializerGenerator } from "./serializer";
import { CodegenRegistry } from "./router";
import { InternalSerializerType } from "../type";
import { Scope } from "./scope";

class MapSerializerGenerator extends BaseSerializerGenerator {
  description: MapTypeDescription;

  constructor(description: TypeDescription, builder: CodecBuilder, scope: Scope) {
    super(description, builder, scope);
    this.description = <MapTypeDescription>description;
  }

  private innerMeta() {
    const key = this.description.options.key;
    const value = this.description.options.value;
    return [this.builder.meta(key), this.builder.meta(value)];
  }

  private innerGenerator() {
    const key = this.description.options.key;
    const value = this.description.options.value;

    const KeyGeneratorClass = CodegenRegistry.get(key.type);
    const ValueGeneratorClass = CodegenRegistry.get(value.type);
    if (!KeyGeneratorClass) {
      throw new Error(`${key.type} generator not exists`);
    }
    if (!ValueGeneratorClass) {
      throw new Error(`${value.type} generator not exists`);
    }
    return [new KeyGeneratorClass(key, this.builder, this.scope), new ValueGeneratorClass(value, this.builder, this.scope)];
  }

  writeStmt(accessor: string): string {
    const [keyMeta, valueMeta] = this.innerMeta();
    const [keyGenerator, valueGenerator] = this.innerGenerator();
    const key = this.scope.uniqueName("key");
    const value = this.scope.uniqueName("value");

    return `
            ${this.builder.writer.varUInt32(`${accessor}.size`)}
            ${this.builder.writer.reserve(`${keyMeta.fixedSize + valueMeta.fixedSize} * ${accessor}.size`)};
            for (const [${key}, ${value}] of ${accessor}.entries()) {
                ${keyGenerator.toWriteEmbed(key)}
                ${valueGenerator.toWriteEmbed(value)}
            }
        `;
  }

  readStmt(accessor: (expr: string) => string): string {
    const [keyGenerator, valueGenerator] = this.innerGenerator();
    const key = this.scope.uniqueName("key");
    const value = this.scope.uniqueName("value");

    const result = this.scope.uniqueName("result");
    const idx = this.scope.uniqueName("idx");
    const len = this.scope.uniqueName("len");

    return `
            const ${result} = new Map();
            ${this.pushReadRefStmt(result)};
            const ${len} = ${this.builder.reader.varUInt32()};
            for (let ${idx} = 0; ${idx} < ${len}; ${idx}++) {
                let ${key};
                let ${value};
                ${keyGenerator.toReadEmbed(x => `${key} = ${x};`)}
                ${valueGenerator.toReadEmbed(x => `${value} = ${x};`)}
                ${result}.set(${key}, ${value});
            }
            ${accessor(result)}
         `;
  }
}

CodegenRegistry.register(InternalSerializerType.MAP, MapSerializerGenerator);
