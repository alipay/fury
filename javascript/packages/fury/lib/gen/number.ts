/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { TypeDescription } from "../description";
import { Builder } from "./builder";
import { BaseSerializerGenerator } from "./serializer";
import { Register } from "./router";
import { InternalSerializerType } from "../type";
import { Scope } from "./scope";

function buildNumberSerializer(writeFun: (builder: Builder, accessor: string) => string, read: (builder: Builder) => string) {
  return class NumberSerializerGenerator extends BaseSerializerGenerator {
    description: TypeDescription;

    constructor(description: TypeDescription, builder: Builder, scope: Scope) {
      super(description, builder, scope);
      this.description = description;
    }

    writeStmt(accessor: string): string {
      return writeFun(this.builder, accessor);
    }

    readStmt(accessor: (expr: string) => string): string {
      return accessor(read(this.builder));
    }
  };
}

Register.reg(InternalSerializerType.UINT8,
  buildNumberSerializer(
    (builder, accessor) => builder.writer.uint8(accessor),
    builder => builder.reader.uint8()
  )
);
Register.reg(InternalSerializerType.INT8,
  buildNumberSerializer(
    (builder, accessor) => builder.writer.int8(accessor),
    builder => builder.reader.int8()
  )
);
Register.reg(InternalSerializerType.UINT16,
  buildNumberSerializer(
    (builder, accessor) => builder.writer.uint16(accessor),
    builder => builder.reader.uint16()
  )
);
Register.reg(InternalSerializerType.INT16,
  buildNumberSerializer(
    (builder, accessor) => builder.writer.int16(accessor),
    builder => builder.reader.int16()
  )
);
Register.reg(InternalSerializerType.UINT32,
  buildNumberSerializer(
    (builder, accessor) => builder.writer.uint32(accessor),
    builder => builder.reader.uint32()
  )
);
Register.reg(InternalSerializerType.INT32,
  buildNumberSerializer(
    (builder, accessor) => builder.writer.int32(accessor),
    builder => builder.reader.int32()
  )
);
Register.reg(InternalSerializerType.UINT64,
  buildNumberSerializer(
    (builder, accessor) => builder.writer.varUInt64(accessor),
    builder => builder.reader.varUInt64()
  )
);
Register.reg(InternalSerializerType.INT64,
  buildNumberSerializer(
    (builder, accessor) => builder.writer.sliLong(accessor),
    builder => builder.reader.sliLong()
  )
);
Register.reg(InternalSerializerType.FLOAT,
  buildNumberSerializer(
    (builder, accessor) => builder.writer.float(accessor),
    builder => builder.reader.float()
  )
);
Register.reg(InternalSerializerType.DOUBLE,
  buildNumberSerializer(
    (builder, accessor) => builder.writer.double(accessor),
    builder => builder.reader.double()
  )
);
