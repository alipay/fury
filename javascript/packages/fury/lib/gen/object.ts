/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { InternalSerializerType, MaxInt32 } from "../type";
import { Scope } from "./scope";
import { CodecBuilder } from "./builder";
import { ObjectTypeDescription, TypeDescription } from "../description";
import { fromString } from "../platformBuffer";
import { CodegenRegistry } from "./router";
import { BaseSerializerGenerator } from "./serializer";

function computeFieldHash(hash: number, id: number): number {
  let newHash = (hash) * 31 + (id);
  while (newHash >= MaxInt32) {
    newHash = Math.floor(newHash / 7);
  }
  return newHash;
}

const computeStringHash = (str: string) => {
  const bytes = fromString(str);
  let hash = 17;
  bytes.forEach((b) => {
    hash = hash * 31 + b;
    while (hash >= MaxInt32) {
      hash = Math.floor(hash / 7);
    }
  });
  return hash;
};

const computeStructHash = (description: TypeDescription) => {
  let hash = 17;
  for (const [, value] of Object.entries((<ObjectTypeDescription>description).options.props).sort()) {
    let id = value.type;
    if (value.type === InternalSerializerType.ARRAY || value.type === InternalSerializerType.TUPLE || value.type === InternalSerializerType.MAP) {
      id = Math.floor(value.type); // TODO add map key&value type into schema hash
    } else if (value.type === InternalSerializerType.FURY_TYPE_TAG) {
      id = computeStringHash((<ObjectTypeDescription>value).options.tag);
    }
    hash = computeFieldHash(hash, id);
  }
  return hash;
};

class ObjectSerializerGenerator extends BaseSerializerGenerator {
  description: ObjectTypeDescription;

  constructor(description: TypeDescription, builder: CodecBuilder, scope: Scope) {
    super(description, builder, scope);
    this.description = <ObjectTypeDescription>description;
  }

  writeStmt(accessor: string): string {
    const options = this.description.options;
    const expectHash = computeStructHash(this.description);
    const tagWriter = this.scope.declare("tagWriter", `${this.builder.classResolver.createTagWriter(this.safeTag())}`);

    return `
            ${tagWriter}.write(${this.builder.writer.ownName()});
            ${this.builder.writer.int32(expectHash)};
            ${Object.entries(options.props).sort().map(([key, inner]) => {
            const InnerGeneratorClass = CodegenRegistry.get(inner.type);
            if (!InnerGeneratorClass) {
                throw new Error(`${inner.type} generator not exists`);
            }
            const innerGenerator = new InnerGeneratorClass(inner, this.builder, this.scope);
            return innerGenerator.toWriteEmbed(`${accessor}${CodecBuilder.safePropAccessor(key)}`);
        }).join(";\n")
            }
        `;
  }

  readStmt(accessor: (expr: string) => string): string {
    const options = this.description.options;
    const expectHash = computeStructHash(this.description);
    const result = this.scope.uniqueName("result");
    return `
        if (${this.builder.reader.int32()} !== ${expectHash}) {
            throw new Error("validate hash failed: ${this.safeTag()}. expect ${expectHash}");
        }
        const ${result} = {
            ${Object.entries(options.props).sort().map(([key]) => {
            return `${CodecBuilder.safePropName(key)}: null`;
        }).join(",\n")}
        };
        ${this.pushReadRefStmt(result)}
        ${Object.entries(options.props).sort().map(([key, inner]) => {
            const InnerGeneratorClass = CodegenRegistry.get(inner.type);
            if (!InnerGeneratorClass) {
                throw new Error(`${inner.type} generator not exists`);
            }
            const innerGenerator = new InnerGeneratorClass(inner, this.builder, this.scope);
            return innerGenerator.toReadEmbed(expr => `${result}${CodecBuilder.safePropAccessor(key)} = ${expr}`);
        }).join(";\n")
            }
        ${accessor(result)}
         `;
  }

  safeTag() {
    return CodecBuilder.replaceBackslashAndQuote(this.description.options.tag);
  }

  toReadEmbed(accessor: (expr: string) => string): string {
    const name = this.scope.declare(
      "tag_ser",
            `fury.classResolver.getSerializerByTag("${this.safeTag()}")`
    );
    return accessor(`${name}.read()`);
  }

  toWriteEmbed(accessor: string): string {
    const name = this.scope.declare(
      "tag_ser",
            `fury.classResolver.getSerializerByTag("${this.safeTag()}")`
    );
    return `${name}.write(${accessor})`;
  }
}

CodegenRegistry.register(InternalSerializerType.FURY_TYPE_TAG, ObjectSerializerGenerator);
