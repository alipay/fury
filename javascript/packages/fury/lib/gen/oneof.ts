/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { OneofTypeDescription, Type, TypeDescription } from "../description";
import { CodecBuilder } from "./builder";
import { BaseSerializerGenerator } from "./serializer";
import { CodegenRegistry } from "./router";
import { InternalSerializerType, RefFlags } from "../type";
import { Scope } from "./scope";

class OneofSerializerGenerator extends BaseSerializerGenerator {
  description: OneofTypeDescription;

  constructor(description: TypeDescription, builder: CodecBuilder, scope: Scope) {
    super(description, builder, scope);
    this.description = <OneofTypeDescription>description;
  }

  writeStmt(): string {
    throw new Error("Type oneof writeStmt can't inline");
  }

  readStmt(): string {
    throw new Error("Type oneof readStmt can't inline");
  }

  toWriteEmbed(accessor: string, excludeHead = false) {
    if (excludeHead) {
      throw new Error("Oneof can't excludeHead");
    }
    if (Object.values(this.description.options.inner).length < 1) {
      throw new Error("Type oneof must contain at least one field");
    }
    const stmts = [
            `if (${accessor} === null || ${accessor} === undefined) {
                ${this.builder.writer.int8(RefFlags.NullFlag)};
            }`,
    ];
    Object.entries(this.description.options.inner).forEach(([key, value]) => {
      const InnerGeneratorClass = CodegenRegistry.get(value.type);
      if (!InnerGeneratorClass) {
        throw new Error(`${value.type} generator not exists`);
      }
      const innerGenerator = new InnerGeneratorClass(value, this.builder, this.scope);
      stmts.push(` if (${CodecBuilder.safeString(key)} in ${accessor}) {
                ${innerGenerator.toWriteEmbed(`${accessor}${CodecBuilder.safePropAccessor(key)}`)}
            }`);
    });
    stmts.push(`
            {
                ${this.builder.writer.int8(RefFlags.NullFlag)};
            }
        `);
    return stmts.join("else");
  }

  toReadEmbed(accessor: (expr: string) => string, excludeHead = false): string {
    const AnyGeneratorClass = CodegenRegistry.get(InternalSerializerType.ANY)!;
    const anyGenerator = new AnyGeneratorClass(Type.any(), this.builder, this.scope);
    return anyGenerator.toReadEmbed(accessor, excludeHead);
  }

  toSerializer() {
    this.scope.assertNameNotDuplicate("read");
    this.scope.assertNameNotDuplicate("readInner");
    this.scope.assertNameNotDuplicate("write");
    this.scope.assertNameNotDuplicate("writeInner");

    const declare = `
          const readInner = (fromRef) => {
            throw new Error("Type oneof readInner can't call directly");
          };
          const read = () => {
            ${this.toReadEmbed(expr => `return ${expr}`)}
          };
          const writeInner = (v) => {
            throw new Error("Type oneof writeInner can't call directly");
          };
          const write = (v) => {
            ${this.toWriteEmbed("v")}
          };
        `;
    return `
            return function (fury, external) {
                ${this.scope.generate()}
                ${declare}
                return {
                  read,
                  readInner,
                  write,
                  writeInner,
                  meta: ${JSON.stringify(this.builder.meta(this.description))}
                };
            }
            `;
  }
}

CodegenRegistry.register(InternalSerializerType.ONEOF, OneofSerializerGenerator);
