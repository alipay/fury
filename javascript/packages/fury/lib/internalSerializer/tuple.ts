/*
 * Copyright 2023 The Fury Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { InternalSerializerType, Serializer } from "../type";
import { Fury } from "../type";

export const tupleSerializer = (fury: Fury, serializers: Serializer[]) => {
  const { binaryReader, binaryWriter, referenceResolver } = fury;

  const { pushReadObject } = referenceResolver;
  const { varInt32: writeVarInt32, reserve: reserves } = binaryWriter;
  const { varInt32: readVarInt32 } = binaryReader;

  return {
    ...referenceResolver.deref(() => {
      const len = readVarInt32();
      const result = new Array(len);
      pushReadObject(result);
      for (let i = 0; i < len; i++) {
        const item = serializers[i];
        result[i] = item.read();
      }
      return result;
    }),
    write: referenceResolver.withNullableOrRefWriter(InternalSerializerType.TUPLE, (v: any[]) => {
      writeVarInt32(serializers.length);

      for (let i = 0; i < serializers.length; i++) {
        const item = serializers[i];
        reserves(item.config().reserve);
        item.write(v[i]);
      }
    }),
    config: () => {
      return {
        reserve: 7,
      };
    },
  };
};
