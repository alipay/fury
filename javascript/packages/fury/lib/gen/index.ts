/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { InternalSerializerType, Fury } from "../type";
import { ArrayTypeDescription, MapTypeDescription, ObjectTypeDescription, SetTypeDescription, TupleTypeDescription, TypeDescription } from "../description";
import { CodegenRegistry } from "./router";
import { CodecBuilder } from "./builder";
import { Scope } from "./scope";
import "./array";
import "./object";
import "./string";
import "./binary";
import "./bool";
import "./datetime";
import "./map";
import "./number";
import "./set";
import "./any";
import "./tuple";
import "./typedArray";

export const generate = (fury: Fury, description: TypeDescription) => {
  const InnerGeneratorClass = CodegenRegistry.get(description.type);
  if (!InnerGeneratorClass) {
    throw new Error(`${description.type} generator not exists`);
  }
  const scope = new Scope();
  const generator = new InnerGeneratorClass(description, new CodecBuilder(scope, fury), scope);

  const funcString = generator.toSerializer();
  const afterCodeGenerated = fury.config?.hooks?.afterCodeGenerated;
  if (typeof afterCodeGenerated === "function") {
    return new Function(afterCodeGenerated(funcString));
  }
  return new Function(funcString);
};

function regDependencies(fury: Fury, description: TypeDescription) {
  if (description.type === InternalSerializerType.FURY_TYPE_TAG) {
    const options = (<ObjectTypeDescription>description).options;
    if (options.props) {
      fury.classResolver.registerSerializerByTag(options.tag);
      Object.values(options.props).forEach((x) => {
        regDependencies(fury, x);
      });
      const func = generate(fury, description);
      fury.classResolver.registerSerializerByTag(options.tag, func()(fury, {}));
    }
  }
  if (description.type === InternalSerializerType.ARRAY) {
    regDependencies(fury, (<ArrayTypeDescription>description).options.inner);
  }
  if (description.type === InternalSerializerType.FURY_SET) {
    regDependencies(fury, (<SetTypeDescription>description).options.key);
  }
  if (description.type === InternalSerializerType.MAP) {
    regDependencies(fury, (<MapTypeDescription>description).options.key);
    regDependencies(fury, (<MapTypeDescription>description).options.value);
  }
  if (description.type === InternalSerializerType.TUPLE) {
    (<TupleTypeDescription>description).options.inner.forEach((x) => {
      regDependencies(fury, x);
    });
  }
}

export const generateSerializer = (fury: Fury, description: TypeDescription) => {
  regDependencies(fury, description);
  if (description.type === InternalSerializerType.FURY_TYPE_TAG) {
    return fury.classResolver.getSerializerByTag((<ObjectTypeDescription>description).options.tag);
  }
  const func = generate(fury, description);
  return func()(fury, {});
};
