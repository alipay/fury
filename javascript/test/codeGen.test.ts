/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { describe, expect, test } from '@jest/globals';
import { tupleObjectDescription, tupleObjectType3Description } from './fixtures/tuple';
import { generate } from '../packages/fury/lib/gen/index';
import FuryInternal from '../packages/fury/lib/fury';
import * as beautify from 'js-beautify';

describe('codeGen', () => {
  test('can generate tuple declaration code', () => {
    const fury = FuryInternal({
      refTracking: true, hooks: {
        afterCodeGenerated: (code: string) => {
          return beautify.js(code, { indent_size: 2, space_in_empty_paren: true, indent_empty_lines: true });
        }
      }
    });
    const fn = generate(fury, tupleObjectDescription);
    expect(fn.toString()).toMatchSnapshot();

    const fn2 = generate(fury, tupleObjectType3Description);
    expect(fn2.toString()).toMatchSnapshot();
  })
})
