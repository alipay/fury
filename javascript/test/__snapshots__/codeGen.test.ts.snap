// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codeGen can generate tuple declaration code 1`] = `
"function anonymous(
) {

return function (fury, scope) {
    const { referenceResolver, binaryWriter, classResolver, binaryReader } = fury;
    const { writeNullOrRef, pushReadObject } = referenceResolver;
    const { RefFlags, InternalSerializerType, arraySerializer, tupleSerializer, mapSerializer, setSerializer } = scope;
    
    const tag_0 = classResolver.getSerializerByTag("example.foo.1")
    const tag_1 = classResolver.getSerializerByTag("example.foo.2")
    const tuple_tag_0_tag_1 = tupleSerializer(fury, [tag_0, tag_1])
    const tag_5 = classResolver.getSerializerByTag("example.bar.1")
    const tag_6 = classResolver.getSerializerByTag("example.bar.2")
    const tuple_tag_0_tag_5_tag_6 = tupleSerializer(fury, [tag_0, tag_5, tag_6])
    const tagBuffer = classResolver.tagToBuffer("tuple-object-wrapper");
    const bufferLen = tagBuffer.byteLength;

    const reserves = tag_0.config().reserve + tag_1.config().reserve + tuple_tag_0_tag_1.config().reserve + tag_5.config().reserve + tag_6.config().reserve + tuple_tag_0_tag_5_tag_6.config().reserve;
    return {
        ...referenceResolver.deref(() => {
            const hash = binaryReader.int32();
            if (hash !== 16469457) {
                throw new Error("validate hash failed: tuple-object-wrapper. expect 16469457, but got" + hash);
            }
            {
                
    // relation tag: tuple-object-wrapper
    const result = {
        tuple1: null,
tuple1_: null,
tuple2: null,
tuple2_: null
    };
    pushReadObject(result);
    result.tuple1 = tuple_tag_0_tag_1.read();
result.tuple1_ = tuple_tag_0_tag_1.read();
result.tuple2 = tuple_tag_0_tag_5_tag_6.read();
result.tuple2_ = tuple_tag_0_tag_5_tag_6.read()
    return result;

            }
        }),
        write: referenceResolver.withNullableOrRefWriter(InternalSerializerType.FURY_TYPE_TAG, (v) => {
            classResolver.writeTag(binaryWriter, "tuple-object-wrapper", tagBuffer, bufferLen);
            binaryWriter.int32(16469457);
            binaryWriter.reserve(reserves);
            tuple_tag_0_tag_1.write(v.tuple1);
tuple_tag_0_tag_1.write(v.tuple1_);
tuple_tag_0_tag_5_tag_6.write(v.tuple2);
tuple_tag_0_tag_5_tag_6.write(v.tuple2_)
        }),
        config() {
            return {
                reserve: bufferLen + 8,
            }
        }
    }
}

}"
`;

exports[`codeGen can generate tuple declaration code 2`] = `
"function anonymous(
) {

return function (fury, scope) {
    const { referenceResolver, binaryWriter, classResolver, binaryReader } = fury;
    const { writeNullOrRef, pushReadObject } = referenceResolver;
    const { RefFlags, InternalSerializerType, arraySerializer, tupleSerializer, mapSerializer, setSerializer } = scope;
    
    const type_13 = classResolver.getSerializerById(13)
    const type_1 = classResolver.getSerializerById(1)
    const type_6 = classResolver.getSerializerById(6)
    const type_14 = classResolver.getSerializerById(14)
    const tuple_type_14 = tupleSerializer(fury, [type_14])
    const tuple_type_13_type_1_type_6_tuple_type_14 = tupleSerializer(fury, [type_13, type_1, type_6, tuple_type_14])
    const tagBuffer = classResolver.tagToBuffer("tuple-object-type3-tag");
    const bufferLen = tagBuffer.byteLength;

    const reserves = type_13.config().reserve + type_1.config().reserve + type_6.config().reserve + type_14.config().reserve + tuple_type_14.config().reserve + tuple_type_13_type_1_type_6_tuple_type_14.config().reserve;
    return {
        ...referenceResolver.deref(() => {
            const hash = binaryReader.int32();
            if (hash !== 552) {
                throw new Error("validate hash failed: tuple-object-type3-tag. expect 552, but got" + hash);
            }
            {
                
    // relation tag: tuple-object-type3-tag
    const result = {
        tuple: null
    };
    pushReadObject(result);
    result.tuple = tuple_type_13_type_1_type_6_tuple_type_14.read()
    return result;

            }
        }),
        write: referenceResolver.withNullableOrRefWriter(InternalSerializerType.FURY_TYPE_TAG, (v) => {
            classResolver.writeTag(binaryWriter, "tuple-object-type3-tag", tagBuffer, bufferLen);
            binaryWriter.int32(552);
            binaryWriter.reserve(reserves);
            tuple_type_13_type_1_type_6_tuple_type_14.write(v.tuple)
        }),
        config() {
            return {
                reserve: bufferLen + 8,
            }
        }
    }
}

}"
`;
